File: 基础语法/class-demo.py
import abc

# 1. 抽象基类：定义接口契约
class Shape(abc.ABC):
    @abc.abstractmethod
    def area(self): ...

# 2. 现代元编程：使用 __init_subclass__ 替代简单元类
# 它可以拦截子类定义，比元类更轻量、易读
class BaseMeta:
    def __init_subclass__(cls, creator="System", **kwargs):
        super().__init_subclass__(**kwargs)
        cls.created_by = creator

# 3. 描述符：属性校验协议
class ValidatedName:
    def __set_name__(self, owner, name):
        self.storage_name = f"_{name}"

    def __get__(self, obj, objtype=None):
        return getattr(obj, self.storage_name) if obj else self

    def __set__(self, obj, value):
        if not isinstance(value, str) or not value.strip():
            raise ValueError("Name must be a non-empty string")
        setattr(obj, self.storage_name, value)

# 4. 多重继承与核心类定义
class Animal:
    def move(self): print("⚡ Moving fast...")

class Person(Animal, BaseMeta, creator="ModernMeta"):
    # 限制属性，优化内存性能
    __slots__ = ("_name", "_age", "__secret", "__dict__")

    name = ValidatedName()

    def __init__(self, name: str, age: int):
        self.name = name       # 触发描述符 __set__
        self._age = age        # 保护成员约定
        self.__secret = "TOP"  # 名称修饰 (Name Mangling)

    # --- 常用装饰器 ---
    @property
    def age(self): return self._age

    @classmethod
    def info(cls): return f"Class: {cls.__name__}, Created by: {cls.created_by}"

    @staticmethod
    def species(): return "Homo Sapiens"

    # --- 魔法方法 (Protocols) ---
    def __repr__(self): return f"Person(name={self.name!r})"
    def __call__(self): print(f"🚀 {self.name} is being called!")
    def __len__(self): return len(self.name)
    def __getitem__(self, i): return self.name[i]
    def __iter__(self): yield from self.name
    def __eq__(self, other):
        return isinstance(other, Person) and self.name == other.name

# --- 动态操作与自省 ---
if __name__ == "__main__":
    p = Person("Ada", 25)

    # 属性操作
    setattr(p, "tag", "AI")         # 动态增 (依赖 __dict__ 在 slots 中)
    val = getattr(p, "tag", None)   # 动态取
    exists = hasattr(p, "tag")      # 检查

    # 协议触发
    p()                             # __call__
    print(f"Index 0: {p[0]}")       # __getitem__
    print(f"Info: {Person.info()}") # @classmethod

    # 自省
    print(f"MRO: {Person.mro()}")   # 继承链
    print(f"Vars: {vars(p)}")       # 实例状态

File: 基础语法/class-demo.js
// 28. 抽象基类 (JS 无原生 ABC，需手动抛错)
class Shape {
  constructor() {
    if (this.constructor === Shape) throw new Error("Abstract class");
  }
  area() {
    throw new Error("Method must be implemented");
  }
}

// 29. 属性描述符/拦截 (JS 通常使用 Proxy 或 defineProperty)
// 30. 元类 (JS 无元类，通常使用 Class Decorators 或 Proxy 代理构造函数)

// 1. 类定义 & 13. 继承 (JS 不支持 27. 多重继承)
class Animal {
  move() {
    console.log("Moving...");
  }
}

class Person extends Animal {
  // 18. 私有属性 (真私有)
  #secret = "shh!";

  // 2. 构造函数
  constructor(name, age) {
    super(); // 14. 必须先调用 super()
    // 3. 实例属性
    this.name = name;
    this._age = age; // 18. 约定私有
  }

  // 4. 实例方法
  greet() {
    super.move();
    return `Hi, I'm ${this.name}`;
  }

  // 12. 字符串表示
  toString() {
    return `Person(${this.name})`;
  }

  // 15. 静态方法 & 16. 类方法 (JS 中 static 即类方法)
  static species() {
    return "Human";
  }
  static info() {
    return `This is a ${this.name}`;
  }

  // 17. 属性装饰器 (Getter/Setter)
  get age() {
    return this._age;
  }
  set age(v) {
    this._age = v;
  }

  // 24. 迭代器 (唯一对应的魔法方法)
  *[Symbol.iterator]() {
    for (let char of this.name) yield char;
  }

  // 19-23. JS 不支持运算符重载 (__len__, __add__, __eq__ 等无对应)
  // 21. 实例调用：JS 实例不可作为函数调用，除非实例本身是 Function
}

// --- 动态操作 (6-11, 25-26) ---
const p = new Person("Ada", 25); // 5. new 关键字
p.job = "Dev"; // 6. 动态增
const job = p.job; // 7. 获取
const hasJob = "job" in p; // 8. 检查
delete p.job; // 9. 删除
console.log(p instanceof Person); // 10. 类型检查
console.log(typeof p); // 11. 获取类型 (返回 'object')
console.log(Object.keys(p)); // 25. 列出属性
console.log(Object.entries(p)); // 26. 属性字典对应项
